src/System.Private.CoreLib/src/System/StartupHookProvider.cs

```diff
@@ -0,0 +1,90 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Runtime.Loader;
+using Internal.IO;
+
+namespace System
+{
+    internal class StartupHookProvider
+    {
+        private static string initializeMethodName = "Initialize";
+
+        // Parse a string specifying a list of assemblies and types
+        // containing a startup hook, and call each hook in turn.
+        private static void ProcessStartupHooks(IntPtr startupHooksVariable)
+        {
+            Debug.Assert(startupHooksVariable != IntPtr.Zero);
+            string[] startupHooks = Marshal.PtrToStringUTF8(startupHooksVariable).Split(new char[] { Path.PathSeparator });
+
+            // Parse startup hook variable
+            List<Tuple<string, string>> startupHooksParsed = new List<Tuple<string, string>>();
+            foreach (string startupHook in startupHooks)
+            {
+                string[] assemblyPathAndTypeName = startupHook.Split(new char[] { '!' });
+                if (assemblyPathAndTypeName.Length != 2)
+                {
+                    throw new ArgumentException(SR.Argument_InvalidStartupHookSyntax);
+                }
+                string assemblyPath = assemblyPathAndTypeName[0];
+                string typeName = assemblyPathAndTypeName[1];
+                if (String.IsNullOrEmpty(assemblyPath) || String.IsNullOrEmpty(typeName))
+                {
+                    throw new ArgumentException(SR.Argument_InvalidStartupHookSyntax);
+                }
+                startupHooksParsed.Add(Tuple.Create(assemblyPath, typeName));
+            }
+
+            // Ensure the startup dlls exist
+            foreach (var startupHook in startupHooksParsed)
```

Comment:
Any reason you wanted to pull out the assembly validation logic out of the loop above?

Reply:
Yes, I think it makes sense for syntax errors to be caught before any missing assemblies are reported. For example, a hook like:
```/path/to/missingAssembly.dll!TypeName;/path/to/Assembly.dll!Weird!Syntax!!```
should throw a syntax error, not a missing assembly error.

Similarly, I think it makes sense for the missing file exception to be thrown before any hooks are executed. If I hadn't done the explicit check here, a similar error would have been thrown in CallStartupHook.

Whereas for missing types, I think it makes sense for those to be caught as the hooks are called one-by-one.